### PR TITLE
Remove paranoid-deleted objects from being tracked in embedded relations

### PR DESCRIPTION
**Pull request by JagdeepSingh**

### CODE REVIEW

**Comment by dblock** [comment]
Needs a rebase too.

**Comment by JagdeepSingh** [comment]
@dblock please review this.

**Comment by dblock** [comment]
This looks good. We actually don't even document what mongoid-history does with Mongoid::Paranoya. This change doesn't only affect embedded objects, but the README seems to imply that. Want to rewrite  this part by maybe adding a general section on support for Mongoid::Paranoya?

**Comment by johnnyshields** [comment]
@dblock the current Mongoid::History code works fine with Mongoid::Paranoia for normal docs. It's only embedded docs which have a problem.

**Comment by johnnyshields** [comment]
@JagdeepSingh can you please add paranoid support for `embeds_one` as well?

**Comment by JagdeepSingh** [comment]
@dblock this implementation doesn't affect the parent document (it works as before). This only excludes soft-deleted `embeds_many` documents from track history.

**Comment by JagdeepSingh** [comment]
@johnnyshields @dblock i will work on the change for `embeds_one` documents as well and see how it looks.

**Comment by JagdeepSingh** [comment]
Also, i am thinking of making it independent of `Mongoid::Paranoia`. We should not bind user to use `mongoid_paranoia` for soft-deletion. If users have implemented their own custom libraries for this, it should work for them too. So, i will allow users to configure `paranoia_field` as well as `paranoia_method` with default values of `deleted_at` and `deleted?`, respectively.

**Comment by johnnyshields** [comment]
@JagdeepSingh how about `paranoia_method` which can take either a Symbol (field/method) or a Proc?

**Comment by johnnyshields** [comment]
@JagdeepSingh on second thought, let's just do `paranoid_field`. Paranoid will in 99% of cases need to be a single database field, because the design of paranoia is such that it filters out objects in the DB query.

**Comment by JagdeepSingh** [comment]
@johnnyshields i have done the changes. @dblock please give it a look now.

File: spec/unit/my_instance_methods_spec.rb [hunk 1]
```diff
@@ -96,6 +96,129 @@
...
+      context 'when embeds_one object blank' do
+        let(:emb_one) { nil }
+
+        it 'should not include embeds_one model key' do
+          expect(subject.keys).to_not include 'emb_one'
+        end
+      end
+
+      describe 'embeds_one' do
+        before(:all) do
```
Comment (dblock) [inline spec/unit/my_instance_methods_spec.rb]: Generally you can do this via `let(:mail_class) { Class.new ... }`.

File: spec/unit/tracker_spec.rb [hunk 2]
```diff
@@ -13,6 +13,141 @@ class MyTracker
...
+            expect(subject['remove']).to be_nil
+          end
+        end
+      end
+    end
+
+    after(:all) do
+      Object.send(:remove_const, :TrackerOne)
+      Object.send(:remove_const, :ModelOne)
+      Object.send(:remove_const, :EmbOne)
```
Comment (dblock) [inline spec/unit/tracker_spec.rb]: the `let` avoids having to do that

**Comment by dblock** [comment]
I think I know what my problem is with this.

The README's section title says "Ignore soft-deleted embedded objects from being tracked" and proposes to set the name of the paranoya field, which somehow only applies to embedded objects? It doesn't let me set different fields for different objects, and doesn't apply to the parent object that I am setting the setting on. 

I think a better way to do this would be something like this:

```ruby
class Cover
  track_history :paranoia_field => :deactivated_at
end

class Cover
  track_history :paranoia_field => :scribbled_at
end

class Book
  include Mongoid::Document
  
  embeds_one :cover
  embeds_many :pages

  track_history :on => [:cover, :pages], :paranoia_field => :deleted_at
end
```

What do you think?

**Comment by JagdeepSingh** [comment]
@dblock calling `track_history` inside embedded class will create separate records for embedded objects, as it does on current master. I understand that we need a way to configure paranoia_field for each relation, but we will have to do that in parent class only.

**Comment by JagdeepSingh** [comment]
@dblock may be we can define a class method like `skip_track_history` which takes `paranoia_field: :deactivated_at` as args. And then call this method in each class where paranaoia_field is different from default.

```
class Book
  include Mongoid::History
  track_history on: [:cover, :pages]
  skip_track_history paranoia_field: :removed_at
end

class Cover
  include Mongoid::History
  skip_track_history paranoia_field: :deactivated_at
end
```

**Comment by johnnyshields** [comment]
@dblock I agree with @JagdeepSingh that we don't want to call the `track_history` macro on the embedded classes, because `track_history`'s primary meaning is to create history trackers.

However, I think @JagdeepSingh's `skip_track_history` idea is also highly confusing.

Three ways I can think to do this:

A) Make a separate macro for the embedded classes

```ruby
class Page
  include Mongoid::History
  history_paranoia_field :removed_at
end
```

B) Include the paranoid field in the embedded field whitelist in the parent class.

```ruby
class Book
  track_history :on => [:cover, {pages: [:title, body, {paranoia_field: :deleted_at}]]
end
```

C) Assume the `Mongoid::Paranoid` gem will always be used and infer the paranoid fields from it. Do not allow other paranoid fields not defined by that specific gem.

**Comment by JagdeepSingh** [comment]
Just to confirm, soft-deleting a parent object, (in our example `book`), is no different than updating the `book`. it will just set the paranoia_field for the object. So, we just need to add `paranoia_field` to the list of `tracked_fields` and it will work fine.

**Comment by JagdeepSingh** [comment]
From above 3 solutions, i think 'B' will complicate the options parsing. Also, the paranoia_field for embedded objects actually belong the the embedded class. So, option 'A' is preferred by me.

With that approach, we can use `paranoia_field` in both types of tracking for embedded objects (1. where audits are nested inside parent 2. where separate audits are created for embedded objects).

@dblock what do you think?

**Comment by dblock** [comment]
I like A as well, but I would call it `history` so we can add other options in the future, so 

```ruby
history :paranoia_field => :deleted_at
```

Another variation that I like less than above, but thought I'd bring up is as follows. 

```ruby
embeds_many :pages, :paranoia_field => :deleted_at
```

The drawback is that the option belongs to the relation and not to the class, but the advantage is that it is "close" to `track_history`.

**Comment by johnnyshields** [comment]
@dblock how about `history_options` or something. Name needs to be clear that it does not enable history tracking on that model.

**Comment by dblock** [comment]
I'm cool with `history_options`, but maybe `history_settings` is more idiomatic, no strong feelings.

**Comment by JagdeepSingh** [comment]
Going with `history_settings`

**Comment by JagdeepSingh** [comment]
@dblock @johnnyshields please have a look at the changes whenever you can. Still working on rspecs and README changes.

File: spec/unit/my_instance_methods_spec.rb (hunk 1 above)
Comment (JagdeepSingh) [inline spec/unit/my_instance_methods_spec.rb]: I used `let` for latest commits wherever there was no need of class name constant

**Comment by JagdeepSingh** [comment]
@dblock i have done my changes.

@johnnyshields FYI

File: README.md [hunk 3]
```diff
@@ -256,6 +256,29 @@ Book.tracked_embeds_many          #=> ["pages"]
 Book.tracked_embeds_many?(:pages) #=> true
 ```
 
+**Skip soft-deleted embedded objects with nested tracking**
```
Comment (dblock) [inline README.md]: Maybe the title is misleading and this is about the paranoia field, not about embedded objects tracking.

**Comment by dblock** [comment]
Excellent work @JagdeepSingh, merging.

**Comment by johnnyshields** [comment]
:+1:

**Pull request merged**